### ISSUE TITLE
Running prefactor on core stations only data

### ISSUE THREAD

**Issue by dirkscholte** [lofar-astron/prefactor#170]
Hi, I am trying to run prefactor on a dataset with core stations only but can't seem to fix an error.

Initially I tried running the dataset without changing the calibration and target parsets (these work for datasets with both CS and RS). This parset did not run on my data, it gives the following error:

2018-03-13 10:13:43 DEBUG   genericpipeline.executable_args: compute.dispatch results job 362: job_duration: 3.50067090988, returncode: 1 
2018-03-13 10:13:43 DEBUG   genericpipeline.executable_args: Adding node_logging_information
2018-03-13 10:13:44 ERROR   genericpipeline.executable_args: A job has failed with returncode 1 and error_tolerance is not set. Bailing out!
2018-03-13 10:13:44 WARNING genericpipeline.executable_args: Note: recipe outputs are not complete
2018-03-13 10:13:44 WARNING genericpipeline.executable_args: recipe executable_args completed with errors
2018-03-13 10:13:44 WARNING genericpipeline: dppp reports failure (using executable_args recipe)
2018-03-13 10:13:44 ERROR   genericpipeline: *******************************************
2018-03-13 10:13:44 ERROR   genericpipeline: Failed pipeline run: Pre-Facet-Calibrator
2018-03-13 10:13:44 ERROR   genericpipeline: Detailed exception information:
2018-03-13 10:13:44 ERROR   genericpipeline: <class 'lofarpipe.support.lofarexceptions.PipelineRecipeFailed'>
2018-03-13 10:13:44 ERROR   genericpipeline: dppp failed
2018-03-13 10:13:44 ERROR   genericpipeline: *******************************************
2018-03-13 10:13:44 ERROR   genericpipeline: LOFAR Pipeline finished unsuccesfully.
2018-03-13 10:13:44 WARNING genericpipeline: recipe genericpipeline completed with errors

I also found the following a bit higher up in the log file:
2018-03-13 10:13:40 WARNING node.hulk.jb.man.ac.uk.executable_args.L520998_SB383_uv.dppp.MS: /lofar/lofar-2_19/SL7/lofar/release/bin/NDPPP stderr: 
**** uncaught exception ****

Backtrace follows:
#0  0x7f826db6decf in LOFAR::Exception::terminate() at Exception.cc:89
#1  0x7f8265948946 in std::rethrow_exception(std::__exception_ptr::exception_ptr) at ??:0
#2  0x7f8265948973 in std::terminate() at ??:0
#3  0x7f8265948b93 in __cxa_throw at ??:0
#4  0x7f826c590a32 in casa::MSSelectionErrorHandler::handleError(casa::MSSelectionError&) at ??:0
#5  0x7f826c58626f in casa::MSSelection::runErrorHandler() at ??:0
#6  0x7f826c586635 in casa::MSSelection::toTableExprNode(casa::MSSelectableTable*) at ??:0
#7  0x7f826c5871cc in casa::MSSelection::toTableExprNode(casa::MeasurementSet const*) at ??:0
#8  0x7f826e254a17 in LOFAR::DPPP::MSReader::MSReader(std::string const&, LOFAR::ParameterSet const&, std::string const&, bool) at MSReader.cc:125
#9  0x7f826e1ebb2a in LOFAR::DPPP::DPRun::makeSteps(LOFAR::ParameterSet const&) at DPRun.cc:269
#10 0x7f826e1ed817 in LOFAR::DPPP::DPRun::execute(std::string const&, int, char**) at DPRun.cc:132
#11 0x404cd8 in main at NDPPP.cc:88
terminate called after throwing an instance of 'casa::MSSelectionError'
  what():  Antenna Expression: No match found for token(s) "RS*"  "RS*"

I then decided to try and make sure prefactor wouldn't look for the RS by changing wherever it said "CS*&; RS*&; CS*&RS*" to just "CS*&". This seems to work okay for the calibrator parset but when calibrating the target it throws an error:

2018-03-13 09:53:59 INFO    genericpipeline.executable_args: recipe executable_args completed
2018-03-13 09:54:00 INFO    genericpipeline: Beginning step expand_sourcedb_target
2018-03-13 09:54:00 INFO    genericpipeline: Beginning step gsmcal_parmmap
2018-03-13 09:54:00 INFO    genericpipeline: Beginning step gsmcal_solve
2018-03-13 09:54:00 ERROR   genericpipeline: *******************************************
2018-03-13 09:54:00 ERROR   genericpipeline: Failed pipeline run: Pre-Facet-Target-CS
2018-03-13 09:54:00 ERROR   genericpipeline: Detailed exception information:
2018-03-13 09:54:00 ERROR   genericpipeline: <type 'exceptions.KeyError'>
2018-03-13 09:54:00 ERROR   genericpipeline: 'mapfile'
2018-03-13 09:54:00 ERROR   genericpipeline: *******************************************
2018-03-13 09:54:00 ERROR   genericpipeline: LOFAR Pipeline finished unsuccesfully.
2018-03-13 09:54:00 WARNING genericpipeline: recipe genericpipeline completed with errors

I can't find a way to work around this error, does anyone have any suggestions or a better way to run prefactor on core stations only data?

If this isn't enough information let me know, I can add the parsets or logfiles if needed!

Kind regards,
Dirk

**Comment by twshimwell** [comment]
Hi Dirk, 

Please do add the parsets and the logfiles.

Thanks

**Comment by dirkscholte** [comment]
Hi,
I have added the parsets and the target logfile (pipeline.log). I could not add the calibrator logfile because it was too big (47 MB) but it completed without errors.

I have also uploaded the plots that have been produced by prefactor.
Thank you for looking into it!

![amp_allsols](https://user-images.githubusercontent.com/35468336/37966651-151fae0e-31c1-11e8-84e9-ac7bca43369e.png)
![cal_phases_polxx_dirpointing](https://user-images.githubusercontent.com/35468336/37966652-15361e6e-31c1-11e8-9a33-0e8fc11582ed.png)
![cal_phases_polyy_dirpointing](https://user-images.githubusercontent.com/35468336/37966653-154ea6dc-31c1-11e8-886f-0874a6752ee1.png)
![dclock_allsols](https://user-images.githubusercontent.com/35468336/37966654-156c88fa-31c1-11e8-9c2e-abf1ebb46d2a.png)
![dtec_allsols](https://user-images.githubusercontent.com/35468336/37966655-1586aec4-31c1-11e8-98a3-af41e837134c.png)
![losoto_clock](https://user-images.githubusercontent.com/35468336/37966656-15a3a3a8-31c1-11e8-90d3-627c29a1e632.png)
![losoto_tec](https://user-images.githubusercontent.com/35468336/37966657-15bac6f0-31c1-11e8-9e9c-5db9098f9c61.png)
![matrix_xx](https://user-images.githubusercontent.com/35468336/37966658-15d22b24-31c1-11e8-8939-7310db37c512.png)
![matrix_yy](https://user-images.githubusercontent.com/35468336/37966659-15eb9e1a-31c1-11e8-9b04-65d25137ba83.png)
![phase_xx_yy_offset](https://user-images.githubusercontent.com/35468336/37966660-16060534-31c1-11e8-9ac0-a496b867ed61.png)


[Pre-Facet-Calibrator-CS.parset.txt](https://github.com/lofar-astron/prefactor/files/1851887/Pre-Facet-Calibrator-CS.parset.txt)
[Pre-Facet-Target-CS.parset.txt](https://github.com/lofar-astron/prefactor/files/1851888/Pre-Facet-Target-CS.parset.txt)
[pipeline.log](https://github.com/lofar-astron/prefactor/files/1851892/pipeline.log)

**Comment by twshimwell** [comment]
Hey,

With your calibrator I think you need to rerun after removing the high frequency subbands (generally best on tonly keep those below about 168MHz). This should make your amplitude solutions look better (getting rid of the spikes at higher frequency) and maybe itll fix the clock solutions (which presently are not good).

**Comment by twshimwell** [comment]
For the target from the log it looks like you give it 10subbands which results in 1 concatenated MSfile at the end. However, only 17% of this file is unflagged so its gets removed (you can change the unflagged fraction in the parset to keep it if desired but 17% unflagged isnt good) from the mapfiles and you end up with an empty list in the mapfile.

Do you have more than 10subbands of target data?

**Comment by dirkscholte** [comment]
Hi,
I ran the calibrator pipeline again with only the subbands below 168MHz and ran the target calibration on the full bandwidth. All the concatenated MSfiles have an unflagged fraction of below 50%, so they were all removed. 

I don't think it should be flagging this much of the data, when calibrating the data without prefactor I don't remove nearly as much data.

Below are the output plots of the calibrator pipeline. The clock solutions still look pretty much the same but the spikes are gone in the amplitude solutions.
![amp_allsols](https://user-images.githubusercontent.com/35468336/38031237-afb3f476-3292-11e8-97d8-843bfe964aa6.png)
![cal_phases_polxx_dirpointing](https://user-images.githubusercontent.com/35468336/38031238-afcce0ee-3292-11e8-8bb2-69bfd1139cf9.png)
![cal_phases_polyy_dirpointing](https://user-images.githubusercontent.com/35468336/38031239-afe4166a-3292-11e8-92dc-fd426dbe1ee9.png)
![dclock_allsols](https://user-images.githubusercontent.com/35468336/38031240-affae3a4-3292-11e8-890c-7152e84fe3fd.png)
![dtec_allsols](https://user-images.githubusercontent.com/35468336/38031241-b0125ce6-3292-11e8-8661-db02b8346de7.png)
![losoto_clock](https://user-images.githubusercontent.com/35468336/38031242-b0547e0a-3292-11e8-941f-af8426624dc5.png)
![losoto_tec](https://user-images.githubusercontent.com/35468336/38031243-b06ae7f8-3292-11e8-8f40-3219e36e413c.png)
![matrix_xx](https://user-images.githubusercontent.com/35468336/38031244-b081a0b0-3292-11e8-9576-659ea96d5699.png)
![matrix_yy](https://user-images.githubusercontent.com/35468336/38031245-b0a4ec00-3292-11e8-9f80-3b69fd89cce6.png)
![phase_xx_yy_offset](https://user-images.githubusercontent.com/35468336/38031246-b0ba6260-3292-11e8-90dc-1d8d2b30c4f4.png)

**Comment by twshimwell** [comment]
Maybe it's because you have removed the remote stations which adds a lot to the amount of data flagged? Perhaps either flag the remote stations after prefactor or change the restriction on the amount of data that can be flagged.

**Comment by dirkscholte** [comment]
My data is from observations (the dual imaging+beam-forming mode) with core stations only. Without removing the pointers to the remote stations in the parset file the calibration crashes very early on because it cannot find the remote stations in my data.

Do you know if is there a good way of dealing with this type of data in prefactor?

**Comment by twshimwell** [comment]
Ah yes sorry I forgot that. 

Perhaps try to see which step is doing all the flagging? (Sometimes the ateam bit can do a lot)

**Comment by dirkscholte** [comment]
I have been browsing through the log file but I have to say that I haven't been able to identify the step doing the flagging.  Are there any key words I should be looking out for?

**Comment by twshimwell** [comment]
looks like it might be the ateam clipper part.

chan 0 : 83.25596% output XX flagged
chan 0 : 83.25596% output YY flagged
chan 1 : 83.31292% output XX flagged
chan 1 : 83.31292% output YY flagged
chan 2 : 83.34945% output XX flagged
chan 2 : 83.34945% output YY flagged
chan 3 : 83.43273% output XX flagged
chan 3 : 83.43273% output YY flagged
Total : 83.33777% output XX flagged
Total : 83.33777% output YY flagged

**Comment by dirkscholte** [comment]
Oh yes that definitely looks like it! How do you think I should address this issue. Should I adjust the clip level from 5 Jy to something higher?

**Comment by dirkscholte** [comment]
I have started a run on 10 subbands with a 10 Jy clip level so I'll see how that goes! I'll let you know the outcome when it's finished.

**Comment by dirkscholte** [comment]
The 10 Jy clip level has reduced the percentage of flagged data a bit. I will try higher clip levels to get lower percentages (aiming around 20%). 

10 Jy clip level:
chan 0 : 68.22964% output XX flagged
chan 0 : 68.22964% output YY flagged
chan 1 : 68.32312% output XX flagged
chan 1 : 68.32312% output YY flagged
chan 2 : 68.49452% output XX flagged
chan 2 : 68.49452% output YY flagged
chan 3 : 68.62132% output XX flagged
chan 3 : 68.62132% output YY flagged
Total : 68.41715% output XX flagged
Total : 68.41715% output YY flagged

**Comment by twshimwell** [comment]
Hmm looks like you have quite severe contamination. Did you demix any sources? Or perhaps you are able to demix the archived data before it goes into prefactor? (What is your time and freq resolution?)

**Comment by dirkscholte** [comment]
I didn't demix any sources. But maybe I should... 
The time and frequency resolutions are 10 s and 48.828kHs

**Comment by twshimwell** [comment]
Hmm thats quite course resolution but you could try i guess. Perhaps look at which ateam sources are causing the flagging and where those sources are with respect to the field.

**Comment by dirkscholte** [comment]
Hi, I asked around about the data a bit yesterday and it has been demixed already before we got it here! I have also had a look at the ateam clipper flagging percentage on another CS only dataset. The percentages are similar so I wonder if it might be the case that these flagging percentages somehow are due to the fact that we're working with CS only data.

I imaged the data produced with a higher clip level and by eye it seems to look okay!

I will have a play with the clipping parameters and see if I can see any problems in my image but I think this might just have solved the problem. Thank you very much for your help and patience with me!
